### ISSUE TITLE
Neatest way to re-use validator at attribute-set time?

### ISSUE THREAD

**Issue by SammyTheSnake** [python-attrs/attrs#109]
I'm pretty new to python (software engineer in another language[0]) and I'm trying to use `attr.s` to do what it says on the tin, reduce boilerplate. If I'm making a horrible n00b mistake, sorry!

I'm trying to create an object with validation on both object creation (or rather initialisation) and subsequent setting and I can't work out a neat way to do it. As far as I see, the best option I have is something like this

    def validator_with_long_descriptive_non_namespaced_name(instance, attribute, value):
        # do something clever
    
    @attr.s(
        these={
            "_a": attr.ib(validator=validator_with_long_descriptive_non_namespaced_name)
            # further attributes here
        }
    )
    class SomeClass():
        @property
        def a(self):
            return self._a
        @a.setter(self, value):
            validator_with_long_descriptive_non_namespaced_name(self, '_a', value)
            self._a = value

Which has a couple of drawbacks.
* the validator function is outside of the class's namespace, which might make sense if it's used in multiple classes, but in my case it isn't - I'm checking the value is compatible with another attribute of the same object
* I'm stuck using a hybrid between the `@property` approach and `attr.s` in a purer form
* the name of the attribute in the `__repr__` doesn't match how it appears in code (`some_object.a = 'whatever'` produces `SomeClass(_a='whatever')`
* it ends up being somewhat boilerplatey

Ideally, I'd be able to do something like the following

    @attr.s
    class SomeClass2():
        a = attr.ib(validate_on_set=True)

        @attr.a.validator
        def a(self, value):
            # do something clever

This would mean I get to have my cake and eat it :-)

So, if I'm just missing the "right" way to do this, please let me know. If this is a feature nobody else has a hankering for, then presumably I'm missing something pretty significant, so a pointer on that front would be appreciated, too! Alternatively, if this means a new feature, then can this be a feature request, please. :-)

Cheers & God bless
  Sam "SammyTheSnake" Penny

[0]if you must know, PHP. I'm so sorry.

**Comment by hynek** [comment]
So the original implementation of the validation code actually also ran the validators on assignment.

I’ve removed it before merging because in dynamic languages like Python there’s just too many ways to circumvent it and I personally prefer to not mutate my objects anyways.  Hence attrs rich support for freezing classes and creating new instances with changed attributes (`assoc`).

Of course you can add such a feature yourself by implementing a ``__setattr__`` method that calls your validators whenever you try to set an attribute.

**Comment by skewty** [comment]
### Example:

```python
    def __setattr__(self, name: str, value: Any) -> None:
        """Call the validator when we set the field (by default it only runs on __init__)"""
        for attribute in [a for a in getattr(self.__class__, '__attrs_attrs__', []) if a.name == name]:
            if attribute.type is not None:
                if isinstance(value, attribute.type) is False:
                    raise TypeError('{}.{} cannot set {} because it is not a {}'.format(
                        self.__class__.__name__, attribute.name, value, attribute.type.__name__))
            if attribute.validator is not None:
                attribute.validator(self, attribute, value)
        super().__setattr__(name, value)
```

It would be nice to be able to turn on / off these checks with a boolean in `attrs()` but it doesn't currently accept `**kwargs` for extra arguments.  Perhaps if you could allow a **kwargs and if `kwargs` is not empty after `__attrs_post_init__` returns then raise an exception?  This would prevent wrongly named arguments from not being detected and allow for feature additions such as this.

```python
def __attrs_post_init__(self, **kwargs):
    setattr(self, '_type_check_setters', kwargs.pop('type_check_setters', False))
    setattr(self, '_validate_setters', kwargs.pop('validate_setters', False))
```

**Comment by pwwang** [comment]
With [`attr_property`](https://github.com/pwwang/attr_property), you can do it easily now.

**Comment by hynek** [comment]
This has been fixed since 20.1.0 and `on_setattr` hooks.
